br/pkg/lightning/backend/external/writer.go

```diff
@@ -304,22 +319,36 @@ type Writer struct {
 // WriteRow implements ingest.Writer.
 func (w *Writer) WriteRow(ctx context.Context, idxKey, idxVal []byte, handle tidbkv.Handle) error {
 	keyAdapter := w.keyAdapter
-	w.batchSize += uint64(len(idxKey) + len(idxVal))
 
 	var rowID []byte
 	if handle != nil {
 		rowID = handle.Encoded()
 	}
-	buf := w.kvBuffer.AllocBytes(keyAdapter.EncodedLen(idxKey, rowID))
-	key := keyAdapter.Encode(buf[:0], idxKey, rowID)
-	val := w.kvBuffer.AddBytes(idxVal)
-
-	w.writeBatch = append(w.writeBatch, common.KvPair{Key: key, Val: val})
-	if w.batchSize >= w.memSizeLimit {
+	encodedKeyLen := keyAdapter.EncodedLen(idxKey, rowID)
+	length := encodedKeyLen + len(idxVal) + lengthBytes*2
+	blockIdx, dataBuf, off, allocated := w.kvBuffer.Alloc(length)
+	if !allocated {
 		if err := w.flushKVs(ctx, false); err != nil {
 			return err
 		}
+		blockIdx, dataBuf, off, allocated = w.kvBuffer.Alloc(length)
+		// we now don't support KV larger than blockSize
```

Comment:
seems we will panic? And why do we need to split `memLimit` into multiple blocks.

Reply:
why panic? the mem-limit might be several GB on add-index, it's too large, and hasn't test, but the performance should be similar.

Reply:
I changed an unit test of memLimit = 16, blockSize = 8, Alloc(9). Alloc found the current block is not enough to hold the data and moved to next block, but does not check the pre-allocated blocks have enough capacity

Reply:
https://github.com/pingcap/tidb/pull/47793/commits/0d842cb3a6d1d2e573e0bfc7e4d9eee32d0fa50c